library/psa_crypto_slot_management.h

```diff
@@ -113,21 +113,23 @@ void psa_wipe_all_key_slots(void);
 psa_status_t psa_get_empty_key_slot(psa_key_id_t *volatile_key_id,
                                     psa_key_slot_t **p_slot);
 
-/** Lock a key slot.
+/** Lock a key slot for reading.
  *
  * This function increments the key slot lock counter by one.
+ * If slot is in a WRITING state, this will fail.
+ * If multi-threading is enabled, the caller must hold the
+ * global key slot mutex.
  *
  * \param[in] slot  The key slot.
  *
  * \retval #PSA_SUCCESS
                The key slot lock counter was incremented.
  * \retval #PSA_ERROR_CORRUPTION_DETECTED
- *             The lock counter already reached its maximum value and was not
- *             increased.
+ *             The slot was in a writing state.
  */
 static inline psa_status_t psa_lock_key_slot(psa_key_slot_t *slot)
 {
-    if (slot->lock_count >= SIZE_MAX) {
+    if (slot->lock_count == SIZE_MAX) {
         return PSA_ERROR_CORRUPTION_DETECTED;
```

Comment:
Why is the return value `PSA_ERROR_CORRUPTION_DETECTED` when we attempt to lock for reading a slot that is locked for writing? Isn't there a better error code? ("write lock held"?)

Reply:
The other error code we are adding is already in [the PSA spec](https://arm-software.github.io/psa-api/status-code/1.0/api/status-codes.html#c.PSA_ERROR_SERVICE_FAILURE), are we allowed to create an entirely new error code here? Also, do we also want to return this error if `psa_unlock_key_slot` fails due to `lock_count` being 0? If so, the name might want to be more general than `PSA_..._LOCKED`

Reply:
CREATING seems a misnomer here, since the slots always exist, they sometimes just don't have any key material in them

Reply:
The state is about the content of the slot, not the slot itself. That's the key object, not the key material.

Reply:
As far as current behaviour goes, I think it is simpler to just have a WRITING state. We then have a simple abstraction property that a slot in WRITING is **always** considered unfilled (["A slot with the correct id but in the WRITING state is considered free."](https://github.com/Mbed-TLS/mbedtls/blob/02ed21a377ec4d9d0945ccf7d9ac3467e1c3ec43/docs/architecture/psa-thread-safety.md#slot-states)). This means that operations which fill or empty a key slot both have linearisation points of unlocking the mutex after setting the state to UNUSED, which happens once the contents of the slot have been written to.

I don't think other threads should ever need to distinguish between whether a key slot is currently being filled or emptied - this wouldn't make sense in a linearised trace of operations.

Reply:
The thing about this WRITING state is that it is _not_ a “writing” state, it's a “creation-or-destruction” state. If we later add an actual writing state, it will be separate from that. So I do object to calling it a “writing” state. I think it's easier for human understanding, and also more robust wrt potential implementation bugs, to have separate states for creation and destructions.

“Creation” and “destruction” aren't the best words either because we use them for keys, whereas here I think “creation” encompasses loading from storage, and maybe “destruction” encompasses freeing a slot that contains a persistent key. Or are those done atomically (i.e. with the global lock held)?

Reply:
> we're using READING and WRITING in the way of reader/writer lockers

Not really, as I explained [earlier](https://github.com/Mbed-TLS/mbedtls/pull/8583#discussion_r1414001772). We have a subset of reader-writer lock states, and because of that we make acquire-lock behave differently (error out instead of waiting when the lock is in the other state). So the “WRITING” state is not a write state, it's a different exclusive state (or rather, at least two exclusive states).

Reply:
To capture my suggestion here, how about if these (very!) internal functions return 0 for success and -1 for error, then callers can do

```
status = psa_acquire_key_slot_read_lock(slot);
return (status == 0) ? PSA_SUCCESS : PSA_ERROR_CORRUPTION_DETECTED;
```

Reply:
Returning an error code feels more natural to me, but I won’t object to making them boolean.

(There is a minor trade-off in flexibility vs code-size and hopefully compilers would be able to optimise away the latter.)